R/odbc-connection.R

```diff
@@ -81,6 +76,52 @@ build_connection_string <- function(.string = NULL, ...) {
   paste0(.string, args_string)
 }
 
+#' Quote special character when connecting
+#'
+#' @description
+#' When connecting to a database using odbc, all the arguments are concatenated
+#' into a single connection string that looks like `name1=value1;name2=value2`.
+#' That means if your value contains `=` or `;` then it needs to be quoted.
+#' Other rules means that it's generally a good idea to quote any text that
+#' contains `{`, `}` or white space.
+#'
+#' This function attempts to quote a string in a way that should hopefully
+#' work for most drivers. Unfortunately, there doesn't seem to be an approach
+#' that works everywhere, but hopefully this function solves most common
+#' issues.
+#'
+#' @export
+#' @param x A string to quote.
+#' @return A quoted string, wrapped in `I()`.
+#' @examples
+#' quote_value("abc")
+#' quote_value("ab'c")
+#'
+#' # Real usage is more likely to look like:
+#' \dontrun{
+#' library(DBI)
+#'
+#' con <- dbConnect(
+#'   odbc::odbc(),
+#'   dsn = "reallycooldatabase"
+#'   password = odbc::quote_value(Sys.getenv("MY_PASSWORD"))
+#' )
+#' }
+quote_value <- function(x) {
+  has_single <- grepl("'", x, fixed = TRUE)
+  has_double <- grepl('"', x, fixed = TRUE)
+
+  if (has_single && has_double) {
```

Comment:
From that page:

> If a keyword contains an equal sign (=), it must be preceded by an additional equal sign to indicate that it is part of the keyword.

🤢. ^^^ is about the key rather than the value, and I haven't seen any keys with the same sort of trickiness, but maybe something to keep on our radars in the future.

Reply:
Yeah, fingers crossed that no driver is crazy enough to use a field name that starts with `=`!